### PR TITLE
Video player controller

### PR DESCRIPTION
**Pull request by sigurdm**

### CODE REVIEW

File: example/lib/main.dart [hunk 1]
```diff
@@ -79,81 +79,176 @@ Widget buildCard(String title) {
...
+        ? new Opacity(
+            opacity: 1.0 - animationController.value,
+            child: new Icon(widget.icon, size: 100.0))
+        : new Container();
+  }
+}
+
 typedef Widget VideoWidgetBuilder(
-    BuildContext context, Future<VideoPlayerId> videoPlayerId);
+    BuildContext context, Future<VideoPlayerController> videoPlayerId);
```
Comment (mravn-google) [inline example/lib/main.dart]: Rename `videoPlayerId` to fx `controller`.

File: example/lib/main.dart [hunk 2]
```diff
@@ -79,81 +79,176 @@ Widget buildCard(String title) {
...
+  PlayerLifeCycle(this.uri, this.videoWidgetBuilder);
 
   @override
   _PlayerLifeCycleState createState() =>
       new _PlayerLifeCycleState(videoWidgetBuilder);
 }
 
 class _PlayerLifeCycleState extends State<PlayerLifeCycle> {
-  Future<VideoPlayerId> video;
+  Future<VideoPlayerController> video;
```
Comment (mravn-google) [inline example/lib/main.dart]: `final`?

File: example/lib/main.dart [hunk 3]
```diff
@@ -79,81 +79,176 @@ Widget buildCard(String title) {
...
   _PlayerLifeCycleState(this.videoWidgetBuilder);
 
   @override
   void initState() {
     super.initState();
-    video = VideoPlayerId.create(
-        'http://www.sample-videos.com/video/mp4/720/big_buck_bunny_720p_20mb.mp4');
+    video = VideoPlayerController.create(widget.uri);
+    video.then((VideoPlayerController controller) async {
+      await controller.setLooping(true);
```
Comment (mravn-google) [inline example/lib/main.dart]: Don't do this if we are already dismounted.

File: example/lib/main.dart [hunk 4]
```diff
@@ -79,81 +79,176 @@ Widget buildCard(String title) {
...
+      controller.play();
+    });
   }
 
   @override
   void dispose() {
-    video.then((VideoPlayerId videoId) {
+    video.then((VideoPlayerController videoId) {
```
Comment (mravn-google) [inline example/lib/main.dart]: Rename `videoId`

File: example/lib/main.dart [hunk 5]
```diff
@@ -166,7 +261,7 @@ class _PlayerLifeCycleState extends State<PlayerLifeCycle> {
 }
 
 class Rotating extends StatelessWidget {
```
Comment (mravn-google) [inline example/lib/main.dart]: `RotatingVideo`?

File: example/lib/main.dart [hunk 6]
```diff
@@ -192,23 +287,26 @@ class Rotating extends StatelessWidget {
...
-                                  default:
-                                    if (snapshot.hasError)
-                                      return new Text(
-                                          'Error: ${snapshot.error}');
-                                    else
-                                      return new VideoPlayer(snapshot.data);
-                                }
-                              })),
+                          child: new RotatingWidget(
+                              new FutureBuilder<VideoPlayerController>(
```
Comment (mravn-google) [inline example/lib/main.dart]: Could we decrease the indentation level here by extracting some nested expressions?

File: example/lib/main.dart [hunk 7]
```diff
@@ -192,23 +287,26 @@ class Rotating extends StatelessWidget {
...
+                                        if (snapshot.hasError)
+                                          return new Text(
+                                              'Error: ${snapshot.error}');
+                                        else
+                                          return new VideoPlayPause(
+                                              snapshot.data);
+                                    }
+                                  })),
                         ),
                         width: 300.0,
```
Comment (mravn-google) [inline example/lib/main.dart]: I think this overflows horizontally on iPods?

File: ios/Classes/VideoPlayerPlugin.m [hunk 8]
```diff
@@ -11,7 +15,7 @@ @implementation FrameUpdater
 - (FrameUpdater*)initWithRegistry:(NSObject<FlutterTextureRegistry>*)registry {
   NSAssert(self, @"super init cannot be nil");
   if (self == nil) return nil;
-  _registry = registry;
+    _registry = registry;
```
Comment (mravn-google) [inline ios/Classes/VideoPlayerPlugin.m]: Remove indentation.

File: ios/Classes/VideoPlayerPlugin.m [hunk 9]
```diff
@@ -20,48 +24,82 @@ - (void)onDisplayLink:(CADisplayLink*)link {
...
+@property(nonatomic, readonly) bool disposed;
+@property(nonatomic, readonly) bool isPlaying;
+@property(nonatomic, readonly) bool isLooping;
+@property(nonatomic, readonly) bool isInitialized;
 - (instancetype)initWithURL:(NSURL*)url
                frameUpdater:(FrameUpdater*)frameUpdater;
 - (void)play;
 - (void)pause;
+- (void)setIsLooping:(bool)isLooping;
+- (void)performPlayingState;
```
Comment (mravn-google) [inline ios/Classes/VideoPlayerPlugin.m]: (As on Android): what does it mean to perform a state?

File: ios/Classes/VideoPlayerPlugin.m [hunk 10]
```diff
@@ -20,48 +24,82 @@ - (void)onDisplayLink:(CADisplayLink*)link {
...
+
 @implementation VideoPlayer
 - (instancetype)initWithURL:(NSURL*)url
                frameUpdater:(FrameUpdater*)frameUpdater {
   self = [super init];
+  _isInitialized = false;
+  _isPlaying = false;
+  _isLooping = false;
+  _disposed = false;
   NSAssert(self, @"super init cannot be nil");
```
Comment (mravn-google) [inline ios/Classes/VideoPlayerPlugin.m]: Move to just after `self` assignment above.

File: ios/Classes/VideoPlayerPlugin.m [hunk 11]
```diff
@@ -74,14 +112,75 @@ - (instancetype)initWithURL:(NSURL*)url
   return self;
 }
 
+- (void)observeValueForKeyPath:(NSString*) path
+                      ofObject:(id)object
+                        change:(NSDictionary*)change
+                       context:(void*)context
+{
```
Comment (mravn-google) [inline ios/Classes/VideoPlayerPlugin.m]: Move to end of previous line. Same on line 121.

File: ios/Classes/VideoPlayerPlugin.m [hunk 12]
```diff
@@ -92,10 +191,31 @@ - (CVPixelBufferRef)copyPixelBuffer {
...
+
+- (FlutterError * _Nullable)onListenWithArguments:(id _Nullable)arguments eventSink:(nonnull FlutterEventSink)events {
+    _eventSink = events;
+    return nil;
+}
+
+- (void)dispose {
+    _disposed = true;
+    [_player replaceCurrentItemWithPlayerItem: nil];
+    [[_player currentItem] removeObserver:self forKeyPath:@"status" context: statusContext];
```
Comment (mravn-google) [inline ios/Classes/VideoPlayerPlugin.m]: Remove space after `context:`

File: ios/Classes/VideoPlayerPlugin.m [hunk 13]
```diff
@@ -92,10 +191,31 @@ - (CVPixelBufferRef)copyPixelBuffer {
...
+    [[_player currentItem] removeObserver:self forKeyPath:@"loadedTimeRanges" context:timeRangeContext];
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+    [_eventChannel setStreamHandler: nil];
```
Comment (mravn-google) [inline ios/Classes/VideoPlayerPlugin.m]: Remove space after `setStreamHandler:`

File: ios/Classes/VideoPlayerPlugin.m [hunk 14]
```diff
@@ -104,20 +224,28 @@ + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
   FlutterMethodChannel* channel = [FlutterMethodChannel
       methodChannelWithName:@"video_player"
             binaryMessenger:[registrar messenger]];
-  VideoPlayerPlugin* instance = [[VideoPlayerPlugin alloc] initWithRegistry:[registrar textures]];
+    VideoPlayerPlugin* instance = [[VideoPlayerPlugin alloc] initWithRegistry:[registrar textures] messenger: [registrar messenger]];
```
Comment (mravn-google) [inline ios/Classes/VideoPlayerPlugin.m]: Remove extra spaces (beginning of line and after `messenger:`).

File: ios/Classes/VideoPlayerPlugin.m [hunk 15]
```diff
@@ -104,20 +224,28 @@ + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
...
   [registrar addMethodCallDelegate:instance channel:channel];
 }
 
-- (instancetype)initWithRegistry:(NSObject<FlutterTextureRegistry>*)registry {
+- (instancetype)initWithRegistry:(NSObject<FlutterTextureRegistry>*)registry
+                       messenger:(NSObject<FlutterBinaryMessenger>*) messenger {
```
Comment (mravn-google) [inline ios/Classes/VideoPlayerPlugin.m]: Remove extra space before `messenger`

File: lib/video_player.dart [hunk 16]
```diff
@@ -2,47 +2,275 @@ import 'dart:async';
...
+  VideoProgressColors colors;
+  ProgressBarPainter(this.value, this.colors);
+
+  bool shouldRepaint(CustomPainter painter) {
+    return true;
+  }
+
+  void paint(Canvas canvas, Size size) {
+    canvas.drawRect(
+        new Rect.fromPoints(
```
Comment (mravn-google) [inline lib/video_player.dart]: Some trailing commas here and below will help `dartfmt` do the Flutteright thing.

File: lib/video_player.dart [hunk 17]
```diff
@@ -2,47 +2,275 @@ import 'dart:async';
...
+
+/// Displays the play/buffering status of the video controlled by [controller].
+class VideoProgressBar extends StatefulWidget {
+  final VideoPlayerController controller;
+  final VideoProgressColors colors;
+
+  VideoProgressBar(this.controller, {VideoProgressColors colors})
+      : colors = colors ?? new VideoProgressColors();
+
+  _VideoProgressBarState createState() {
```
Comment (mravn-google) [inline lib/video_player.dart]: You can use the `=> new _VideoProgressBarState();` form here I think.

File: lib/video_player.dart [hunk 18]
```diff
@@ -2,47 +2,275 @@ import 'dart:async';
...
+                    new Duration(milliseconds: range[0]),
+                    new Duration(milliseconds: range[1])))
+                .toList());
+      }
+    });
+  }
 
-  static Future<VideoPlayerId> create(String dataSource) async {
-    int textureId =
+  static Future<VideoPlayerController> create(String dataSource) async {
```
Comment (mravn-google) [inline lib/video_player.dart]: Rename `create` to make it obvious that `dataSource` is a URL?

File: lib/video_player.dart [hunk 19]
```diff
@@ -2,47 +2,275 @@ import 'dart:async';
...
+    Map response =
         await _channel.invokeMethod('create', {'dataSource': dataSource});
-    return new VideoPlayerId._internal(textureId);
+    int textureId = response["textureId"];
```
Comment (mravn-google) [inline lib/video_player.dart]: Couldn't the response be the `textureId` directly, with no wrapping `Map`?

File: lib/video_player.dart [hunk 20]
```diff
@@ -2,47 +2,275 @@ import 'dart:async';
 
 import 'package:flutter/services.dart';
 import 'package:flutter/widgets.dart';
+import 'package:meta/meta.dart';
 
-const MethodChannel _channel = const MethodChannel('video_player');
+final MethodChannel _channel = const MethodChannel('video_player')
```
Comment (mravn-google) [inline lib/video_player.dart]: `plugins.flutter.io/video_player`

File: lib/video_player.dart [hunk 21]
```diff
@@ -2,47 +2,275 @@ import 'dart:async';
...
+class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
+  final int _textureId;
+  Timer timer;
+  bool isDisposed = false;
 
-  VideoPlayerId._internal(int textureId) : textureId = textureId;
+  VideoPlayerController._internal(int textureId)
+      : _textureId = textureId,
+        super(new VideoPlayerValue(duration: null)) {
+    new EventChannel("videoPlugin/videoEvents$textureId")
```
Comment (mravn-google) [inline lib/video_player.dart]: `plugins.flutter.io/video_player/events$textureId`

File: lib/video_player.dart [hunk 22]
```diff
@@ -2,47 +2,275 @@ import 'dart:async';
...
+import 'package:meta/meta.dart';
 
-const MethodChannel _channel = const MethodChannel('video_player');
+final MethodChannel _channel = const MethodChannel('video_player')
+  // This will clear all open videos on the platform when a full restart is
+  // performed.
+  ..invokeMethod("init");
 
-/// A reference to a platform video player playing into a [Texture].
+class DurationRange {
```
Comment (mravn-google) [inline lib/video_player.dart]: We may want to give this value class a `toString`, `operator=`, and `hashCode`.

File: ios/Classes/VideoPlayerPlugin.m [hunk 23]
```diff
@@ -127,19 +255,36 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
...
+         eventChannelWithName: [NSString stringWithFormat: @"videoPlugin/videoEvents%lld", textureId]
+              binaryMessenger: _messenger];
+    [eventChannel setStreamHandler: player];
+    player.eventChannel = eventChannel;
     _players[@(textureId)] = player;
-    result(@(textureId));
+    result(@{@"textureId": @(textureId)});
   } else {
     NSDictionary* argsMap = call.arguments;
     NSUInteger textureId = ((NSNumber*) argsMap[@"textureId"]).unsignedIntegerValue;
```
Comment (mravn-google) [inline ios/Classes/VideoPlayerPlugin.m]: `int64_t`

File: ios/Classes/VideoPlayerPlugin.m [hunk 24]
```diff
@@ -104,20 +224,28 @@ + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
...
+                       messenger:(NSObject<FlutterBinaryMessenger>*) messenger {
   self = [super init];
   NSAssert(self, @"super init cannot be nil");
   _registry = registry;
+  _messenger = messenger;
   _players = [NSMutableDictionary dictionaryWithCapacity:1];
   return self;
 }
 
 - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
```
Comment (mravn-google) [inline ios/Classes/VideoPlayerPlugin.m]: Remove space after `:` several times in this method.

File: ios/Classes/VideoPlayerPlugin.m [hunk 25]
```diff
@@ -104,20 +224,28 @@ + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
...
-  if ([@"create" isEqualToString:call.method]) {
+  if ([@"init" isEqualToString: call.method]) {
+    for(NSNumber* textureId in _players) {
```
Comment (mravn-google) [inline ios/Classes/VideoPlayerPlugin.m]: Add space after `for`

**Comment by sigurdm** [comment]
Thanks for the review!

**Pull request merged**